Add a high-level description of NSM components (#265)
networkservicemesh/site: content/docs/concepts/components_description.md

+++
title = "NSM components"
weight = 4
summary = "Network Service Mesh (NSM): The Hybrid/Multi-cloud IP Service Mesh"
+++

> ## NSM Services

Reviewer: Perhaps `NSM gRPC Services`? Otherwise I think there is confusion. We use the word `Service` in many contexts

**Proto File**: https://github.com/networkservicemesh/api/blob/main/pkg/api/networkservice/networkservice.proto
```proto
service NetworkService {
    rpc Request(NetworkServiceRequest) returns (connection.Connection);
    rpc Close(connection.Connection) returns (google.protobuf.Empty);
}
```

2. **Registry** is used for registering and unregistering **Network Services** and **Network Service Endpoints** in NSM, as well as for searching.<br/>
**Proto File**: https://github.com/networkservicemesh/api/blob/main/pkg/api/registry/registry.proto
```proto
service NetworkServiceEndpointRegistry {
    rpc Register (NetworkServiceEndpoint) returns (NetworkServiceEndpoint);
    rpc Find (NetworkServiceEndpointQuery) returns (stream NetworkServiceEndpointResponse);
    rpc Unregister (NetworkServiceEndpoint) returns (google.protobuf.Empty);
}

service NetworkServiceRegistry {
    rpc Register (NetworkService) returns (NetworkService);
    rpc Find (NetworkServiceQuery) returns (stream NetworkServiceResponse);
    rpc Unregister (NetworkService) returns (google.protobuf.Empty);
}
```

> 3. **MonitorConnections** service is used for monitoring or searching connections with a specific selector.<br/>

Reviewer: MonitorConnections --> MonitorConnection

**Proto File**: https://github.com/networkservicemesh/api/blob/main/pkg/api/networkservice/connection.proto
```proto
service MonitorConnection {
    rpc MonitorConnections(MonitorScopeSelector) returns (stream ConnectionEvent);
}
```

## Network Service Manager (NSMgr)
**Network Service Manager** is one of the key components of NSM, which is responsible for discovering **Network Services** and **Network Service Endpoints** and processing requests from clients. It must be located on the same machine as the NSM client in order to provide it with a connection to the NSM network. It can also serve as a registry if there is no real registry.

> **Image**: https://github.com/networkservicemesh/cmd-nsmgr/pkgs/container/cmd-nsmgr

Reviewer: Image --> Repository

Reviewer 2: I don't think that it's a repository, but I agree that the image is not clear.

Actually, it's a GitHub package. So let's rename the image to the package.

Thoughts?

**Implements services**:
1. NetworkService
2. Registry
3. MonitorConnection

**Responsibilities**:
1. NetworkService discovery
2. Forwarder discovery
3. Storage for valid connections from clients
4. Provides access to the network namespace of the clients and endpoints for the forwarder
5. Can serve as a registry (optional) 

> ## Forwarder VPP

Reviewer: I would suggest making one **Forwarder** header, and include

-  Forwarder VPP

-  Forwarder OVS

**Forwarder VPP** is one of the datapath providers in NSM. Its main responsibility is to configure network interfaces requested by NSM clients. **Forwarder VPP** uses [**VPP Framework**](https://s3-docs.fd.io/vpp/24.02/) to create the datapath.

**Image**: https://github.com/networkservicemesh/cmd-forwarder-vpp/pkgs/container/cmd-forwarder-vpp

**Implements services**:
1. NetworkService

**Responsibilities**:
1. Creates client-side and endpoint-side interfaces
2. Chooses mechanism types for connections between clients and endpoints
3. Collects stats from interfaces
4. Load-balancing for endpoints beloning to the same **Network Service**

> 5. Fills in the `ConnectionContext` for the connection

Reviewer: 6. Connects with other forwarders located on different nodes or clusters.

```proto
message ConnectionContext {
    IPContext ip_context = 1;
    DNSContext dns_context = 2;
    EthernetContext ethernet_context = 3;
    map<string, string> extra_context = 4;
    uint32 MTU = 5;
}
```

## Forwarder OVS
**Forwarder OVS** is an analogue of **Forwarder VPP**, which uses [**Open vSwitch (OVS)**](https://www.openvswitch.org/) as a backend for creating network interfaces.

**Image**: https://github.com/networkservicemesh/cmd-forwarder-ovs/pkgs/container/cmd-forwarder-ovs

## Registry
**Registry** is used as a storage for **Network Service** and **Network Service Endpoint** entries.

**Image**: https://github.com/networkservicemesh/cmd-registry-k8s/pkgs/container/cmd-registry-k8s

**Implements services**:

> 1.  Registry

Reviewer: 1. NetworkService
2. NetworkServiceEndpointRegistry

3. Proxies local queries to the **Registry** on the cluster

## Network Service Manager Proxy (NSMgr Proxy)
**Network Service Manager Proxy** allows NSM to establish connections between clients and endpoints located in different NSM domains. This component is used for proxing connection requests to other **Network Service Manager Proxies** located in other NSM domains.

**Image**: https://github.com/networkservicemesh/cmd-nsmgr-proxy/pkgs/container/cmd-nsmgr-proxy

**Implements services**:
1. NetworkService
2. MonitorConnection

**Responsibilities**:
1. Swaps endpoint's local IP to domain's external IP
2. Translates **Network Service** and **Network Service Endpoint** names in the connection into names suitable for other NSM domains

## Admission Webhook k8s
Admission Webhook K8s simplifies working with NSM if NSM is deployed in a Kubernetes cluster. This component automatically injects NSM clients into Kubernetes pods with NSM annotations. It can also inject NSM clients into entire pod namespaces.

**Image**: https://github.com/networkservicemesh/cmd-admission-webhook-k8s/pkgs/container/cmd-admission-webhook-k8s

**Responsibilities**:

> 1. Injects NSM Clients into Kubernetes pods or namespaces if they have an NSM annotation

Reviewer: It doesn't add NSM Clients to namespace

**Image**: https://github.com/networkservicemesh/cmd-exclude-prefixes-k8s/pkgs/container/cmd-exclude-prefixes-k8s

**Responsibilities**:
1. Obtains and updates prefixes used in the Kubernetes Cluster

## Map IP K8s
**Map IP K8s** builds and updates a map that transfers internal pods' IPs of the Kubernetes Cluster to the external nodes' IPs. This map is used by **Network Service Manager Proxy** for interdomain NSM scenarios when NSM is deployed in the Kubernetes Cluster.

**Image**: https://github.com/networkservicemesh/cmd-map-ip-k8s/pkgs/container/cmd-map-ip-k8s

**Responsibilities**:
1.  Builds and updates a map that transfers internal pods' IPs to external nodes' IPs

## Network Service Endpoint (NSE)
**Network Service Endpoint** provides NSM clients with access to **Network Services**.

**Image**: https://github.com/networkservicemesh/cmd-nse-icmp-responder/pkgs/container/cmd-nse-icmp-responder

**Implements services**:
1. NetworkService

**Responsibilities**:
1. Provides NSM clients with access to **Network Services**
2. Fills in remaining connection properties like IP, hardware addresses, and DNS configurations for the user's goals.

## Network Service Client (NSC)

> **Network Service Client** allows external workloads to request access to NSM's **Network Services** and maintains connections ([Controlplane](https://networkservicemesh.io/docs/concepts/features/healing) and [Dataplane](https://networkservicemesh.io/docs/concepts/features/datapath_healing) healing). This component can also provide a local DNS Server for accessing NSM resources by their names. Supports only **kernel** and **vfio** mechanisms.

Reviewer: I wouldn't point out that it only supports kernel and vfio, because nsc-memif is also NSC.
We can add a link to NSC-memif below (Image -> Repositories)

**Image**: https://github.com/networkservicemesh/cmd-nsc/pkgs/container/cmd-nsc

**Responsibilities**:
1. Requests connections to **Network Services**
2. Connection healing
3. Can serve as a DNS Server

## External Client

> **External Client** is any workload which wants to connect to NSM's **Network Services**. To achieve it **External Client** uses **Network Service Client** which should be located on the same machine.

Reviewer: on the same machine -> in the same pod as sidecar.

## Network Service Client Init (NSC Init)
**Network Service Client Init** is used as **Init Container** inside Kubernetes pods with **External Clients**. Its main responsibility is to connect to **Network Service** before the main workload starts. Connection created by **NSC Init** is then used and maintained by **NSC**. This component supports only **kernel** and **vfio** mechanisms.

**Image**: https://github.com/networkservicemesh/cmd-nsc-init/pkgs/container/cmd-nsc-init

**Responsibilities**:
1. Requests a connection to **Network Service**

## Network Service Endpoint vl3 (NSE vl3)
**Network Service Endpoint vl3** is used to build virtual L3 networks in NSM using [**VPP Framework**](https://s3-docs.fd.io/vpp/24.02/). vl3 networks are very similar to the standard network in a Kubernetes Cluster. vl3 Endpoints route traffic between clients connected to the vl3 network. They can also be configured to provide DNS for connected clients. 

**Image**: https://github.com/networkservicemesh/cmd-nse-vl3-vpp/pkgs/container/cmd-nse-vl3-vpp

**Implements services**:
1. NetworkService

**Responsibilities**:
1. Builds virtual L3 network

> 2. Routes clients' traffic over the vl3 network

Reviewer: clients

+++
title = "Description of NSM components"
weight = 4
summary = "Network Service Mesh (NSM): The Hybrid/Multi-cloud IP Service Mesh"
+++

# NSM Services

> 1. **NetworkService** - used for requesting and closing connections between clients and endpoints.

Reviewer 2: ```suggestion
1. **NetworkService** is used for requesting and closing network connections between clients and endpoints.
```

Reviewer 2: Also attach a link to the proto file.

```proto
service NetworkService {
    rpc Request(NetworkServiceRequest) returns (connection.Connection);
    rpc Close(connection.Connection) returns (google.protobuf.Empty);
}
```

> 2. **Registry** - used for registering and unregistering **Network Services** and **Network Service Endpoints** in NSM, as well as for searching

Reviewer 2: ```suggestion
2. **Registry** is used for registering and unregistering **Network Services** and **Network Service Endpoints** in NSM, as well as for searching.
```

Reviewer 2: Also attach a link to the proto file.

```proto
service NetworkServiceEndpointRegistry {
    rpc Register (NetworkServiceEndpoint) returns (NetworkServiceEndpoint);
    rpc Find (NetworkServiceEndpointQuery) returns (stream NetworkServiceEndpointResponse);
    rpc Unregister (NetworkServiceEndpoint) returns (google.protobuf.Empty);
}

service NetworkServiceRegistry {
    rpc Register (NetworkService) returns (NetworkService);
    rpc Find (NetworkServiceQuery) returns (stream NetworkServiceResponse);
    rpc Unregister (NetworkService) returns (google.protobuf.Empty);
}
```

> 3. **MonitorConnections** - used for monitoring connectons using specific selectors.

Reviewer 2: ```suggestion
3. **MonitorConnections** service is used for monitoring or searching connections by a specific selector.
```

# Exclude Prefixes K8s
**Exclude Prefixes K8s** collects prefixes from various sources in the Kubernetes Cluster including pods, services and config maps. These prefixes are used by NSM IPAM to avoid the intersection of prefixes of NSM and Kubernetes networks.

**Image**: https://github.com/networkservicemesh/cmd-exclude-prefixes-k8s

**Responsibilities**:
1. Obtains and updates prefixes used in the Kubernetes Cluster

# Map IP K8s
**Map IP K8s** builds and updates a map that transfers internal pods' IPs of the Kubernetes Cluster to the external nodes' IPs. This map is used by **Network Service Manager Proxy** for interdomain NSM scenarios when NSM is deployed in the Kubernetes Cluster.

**Image**: https://github.com/networkservicemesh/cmd-map-ip-k8s

**Responsibilities**:
1.  Builds and updates a map that transfers internal pods' IPs to external nodes' IPs

# Network Service Endpoint (NSE)
**Network Service Endpoint** provides NSM clients with access to **Network Services**.

**Image**: https://github.com/networkservicemesh/cmd-nse-icmp-responder

**Implements**:
1. NetworkService

**Responsibilities**:
1. Provides NSM clients with access to **Network Services**

> 2. Allocates IPs inside the NSM network for NSM clients and for itself

Reviewer 2: ```suggestion
2. fills in remaining connection properties like IP, hardware addresses, and DNS configurations for the user's goals.
```
